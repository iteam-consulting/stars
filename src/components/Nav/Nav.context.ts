import { Record, RecordOf } from 'immutable';
import { createContext } from 'react';

export type NavState = {
  pathname: string;
};

export const NavContextRecordType = Record<NavState>({
  pathname: '/',
});

export const NavContext = createContext<RecordOf<NavState>>(
  NavContextRecordType()
);
