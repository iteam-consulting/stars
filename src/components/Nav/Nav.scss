@import '../variables';

nav {
  display: flex;
  flex-direction: column;
  width: var(--nav-width, $nav-width);
  padding-right: 1rem;
}
