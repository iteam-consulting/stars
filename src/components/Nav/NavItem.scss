@import '../variables';

.nav-item {
  border-radius: 0.4rem;
  font-size: 1.5rem;
  padding-left: 1rem;

  &.active {
    background-color: var(--color-primary-300, $color-primary-300);
  }

  .nav-item-list & {
    padding-left: 2rem;
    display: list-item;

    &::marker {
      margin-left: 0.5rem;
      color: var(--color-text-muted, $color-text-muted);
    }

    &:hover::marker,
    &.active::marker {
      color: var(--color-primary, $color-primary);
    }

    a {
      width: calc(100% - 22px);
    }
  }

  & a {
    color: var(--color-text-muted, $color-text-muted);
    display: inline-block;
    padding: 0.7rem 0;
    text-decoration: none;
    width: 100%;
  }

  &:hover a,
  &.active a {
    color: var(--color-primary, $color-primary);
  }
}
