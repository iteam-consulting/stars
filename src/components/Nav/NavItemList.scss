@import '../variables';

.nav-item-list {
  color: var(--color-text-muted, $color-text-muted);

  .arrow-icon svg path {
    transition: none !important;
  }

  &.active {
    color: var(--color-primary, $color-primary);

    .arrow-icon svg path {
      fill: var(--color-primary, $color-primary) !important;
    }
  }

  &--title {
    align-items: center;
    cursor: pointer;
    display: flex;
    font-size: 1.5rem;
    padding: 0.5rem 0 0 1rem;
    margin-bottom: 0.5rem;

    .arrow-icon {
      margin-left: auto;
    }

    &:hover {
      color: var(--color-primary, $color-primary);

      .arrow-icon svg path {
        fill: var(--color-primary, $color-primary) !important;
      }
    }
  }

  &--items {
    list-style: inside;
  }

  .arrow-icon svg {
    transform: rotateZ(0);
    transition: transform ease-in-out $transition-speed;
  }

  &.show .arrow-icon svg {
    transform: rotateZ(90deg);
  }
}
