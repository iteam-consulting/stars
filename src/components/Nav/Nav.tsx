import { History } from 'history';
import React, { HTMLAttributes, ReactNode, useState } from 'react';
import AccordionProvider from '../Accordion/AccordionProvider';
import { NavContext, NavContextRecordType } from './Nav.context';
import './Nav.scss';

export type NavProps = {
  history: History;
  children?: ReactNode;
} & HTMLAttributes<HTMLDivElement>;

function Nav({ history, children, ...props }: NavProps) {
  const [state, setState] = useState(
    NavContextRecordType({
      pathname: history?.location.pathname,
    })
  );

  React.useEffect(() => {
    return history.listen(state => {
      setState(s => s.set('pathname', state.pathname));
    });
  }, [history, setState]);

  return (
    <NavContext.Provider value={state}>
      <AccordionProvider>
        <nav {...props}>{children}</nav>
      </AccordionProvider>
    </NavContext.Provider>
  );
}

export default Nav;
