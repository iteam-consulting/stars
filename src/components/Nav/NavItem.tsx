import classNames from 'classnames';
import { HTMLAttributes, ReactElement, useContext } from 'react';
import { NavContext } from './Nav.context';

type LinkishProps = {
  to: string;
};

export type NavItemProps = {
  children: ReactElement<LinkishProps>;
} & HTMLAttributes<HTMLDivElement>;

function NavItem({ children, className, ...props }: NavItemProps) {
  const state = useContext(NavContext);
  const classes = classNames('nav-item', className, {
    active: state.pathname?.startsWith(children.props.to),
  });

  return (
    <div className={classes} {...props}>
      {children}
    </div>
  );
}

export default NavItem;
