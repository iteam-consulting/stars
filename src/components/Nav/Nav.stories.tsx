import { Meta, Story } from '@storybook/react';
import Nav, { NavProps } from './Nav';
import { Link, Route, Router, Switch } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import NavItem from './NavItem';
import NavItemList from './NavItemList';

export default {
  title: 'Components/Nav',
  component: Nav,
} as Meta;

type StoryProps = Omit<NavProps, 'history' | 'children'>;
const history = createMemoryHistory();

const Template: Story<StoryProps> = args => {
  return (
    <Router history={history}>
      <Nav {...args} history={history}>
        <NavItem>
          <Link to="/page-1">Page 1</Link>
        </NavItem>
        <NavItem>
          <Link to="/page-2">Page 2</Link>
        </NavItem>
        <NavItemList title="More stuff" path="/more">
          <NavItem>
            <Link to="/more/page-1">Page 1</Link>
          </NavItem>
          <NavItem>
            <Link to="/more/page-2">Page 2</Link>
          </NavItem>
        </NavItemList>
      </Nav>
      <Switch>
        <Route path="/page-1">
          <span>Hello from Page 1</span>
        </Route>
        <Route path="/page-2">
          <span>Hello from Page 2</span>
        </Route>
        <Route path="/more/page-1">
          <span>Hello from more Page 1</span>
        </Route>
        <Route path="/more/page-2">
          <span>Hello from more Page 2</span>
        </Route>
        <Route>
          <span>Not found</span>
        </Route>
      </Switch>
    </Router>
  );
};

export const Default = Template.bind({});
