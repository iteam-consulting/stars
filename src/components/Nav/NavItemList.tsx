import classNames from 'classnames';
import React, {
  ComponentType,
  HTMLAttributes,
  SVGProps,
  useContext,
} from 'react';
import Accordion from '../Accordion/Accordion';
import { NavContext } from './Nav.context';
import { ReactComponent as AngleIcon } from './AngleRight.svg';
import './NavItemList.scss';
import SVGIcon from '../SVGIcon/SVGIcon';
import AccordionProvider from '../Accordion/AccordionProvider';

export type NavItemListProps = {
  icon?: ComponentType<SVGProps<SVGElement>>;
  path: string;
  title: string;
  children?: React.ReactNodeArray;
} & HTMLAttributes<HTMLDivElement>;

function NavItemList({
  children,
  className,
  path,
  title,
  ...props
}: NavItemListProps) {
  const state = useContext(NavContext);

  const classes = classNames('nav-item-list', className, {
    active: state.pathname.startsWith(path),
  });

  return (
    <Accordion {...props} id={path} className={classes}>
      <div className="nav-item-list--title">
        <span>{title}</span>
        <SVGIcon className="arrow-icon" variant="xs">
          <AngleIcon />
        </SVGIcon>
      </div>
      <div className="nav-item-list--items">
        <AccordionProvider>{children}</AccordionProvider>
      </div>
    </Accordion>
  );
}

export default NavItemList;
