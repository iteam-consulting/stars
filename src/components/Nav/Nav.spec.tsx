import { render, act, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Link, Router } from 'react-router-dom';
import Nav from './Nav';
import NavItem from './NavItem';
import NavItemList from './NavItemList';

describe('Nav', () => {
  test('Nav renders properly', () => {
    const { container } = render(<Nav history={createMemoryHistory()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Nav is location aware', async () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <Nav history={history}>
          <NavItemList path="/test" title="Test">
            <NavItem>
              <Link to="/test/hello">Hello</Link>
            </NavItem>
          </NavItemList>
        </Nav>
      </Router>
    );
    const link = screen.getByText(/Hello/i);
    const section = screen.getByTitle(/Test/i);

    expect(link.parentElement!.classList).not.toContain('active');
    expect(section.classList).not.toContain('active');

    act(() => {
      history.push('/test/hello');
    });

    expect(link.parentElement!.classList).toContain('active');
    expect(section.classList).toContain('active');
  });

  test('Nav renders a complex tree properly', () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <Nav history={history}>
          <NavItem>
            <Link to="/test/link1">Link 1</Link>
          </NavItem>
          <NavItemList path="/test" title="Test">
            <NavItem>
              <Link to="/test/hello">Hello</Link>
            </NavItem>
          </NavItemList>
        </Nav>
      </Router>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
