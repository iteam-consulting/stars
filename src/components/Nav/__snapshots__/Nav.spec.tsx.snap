// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Nav renders a complex tree properly 1`] = `
<nav>
  <div
    class="nav-item"
  >
    <a
      href="/test/link1"
    >
      Link 1
    </a>
  </div>
  <div
    class="accordion nav-item-list"
    title="Test"
  >
    <div
      class="nav-item-list--title"
      id="/test"
    >
      <span>
        Test
      </span>
      <span
        class="arrow-icon xs svg-icon"
      >
        <svg>
          AngleRight.svg
        </svg>
      </span>
    </div>
    <div
      class="content nav-item-list--items"
      style="max-height: 0;"
    >
      <div
        class="nav-item"
      >
        <a
          href="/test/hello"
        >
          Hello
        </a>
      </div>
    </div>
  </div>
</nav>
`;

exports[`Nav Nav renders properly 1`] = `<nav />`;
